v2/netconf/server/netconf/server.go

```diff
@@ -0,0 +1,273 @@
+}
+
+// SessionCallback defines the caller supplied callback functions.
+type SessionCallback interface {
+	// Capabilities is called to retrieve the capabilities that should be advertised to the client.
+	// If the callback returns nil, the default set of capabilities is used.
+	Capabilities() []string
+	// HandleRequest is called to handle an RPC request.
+	HandleRequest(req *RpcRequestMessage) *RpcReplyMessage
+}
+
+type SessionFactory func(*SessionHandler) SessionCallback
+
+// SessionHandler represents the server side of an active netconf SSH session.
+type SessionHandler struct {
+
+	// server references the Netconf server that launched the session.
+	server *Server
+
+	// svrcon is the underlying ssh server connection.
+	svrcon *xssh.ServerConn
+
+	// ch is the underlying transport channel.
+	ch xssh.Channel
+
+	// The codecs used to handle client i/o
+	enc *codec.Encoder
+	dec *codec.Decoder
+
+	// Serialises access to encoder (avoiding contention between sending notifications and request responses).
+	encLock sync.Mutex
+
+	// The capabilities advertised to the client.
+	capabilities []string
+	// The session id to be reported to the client.
+	sid uint64
+
+	// Channel used to signal successful receipt of client capabilities.
+	hellochan chan bool
+
+	// The HelloMessage sent by the connecting client.
+	ClientHello *common.HelloMessage
+
+	// Caller supplied callbacks
+	cb SessionCallback
+}
+
+// RpcRequestMessage and rpcRequest represent an RPC request from a client, where the element type of the
+// request body is unknown.
+type RpcRequestMessage struct {
+	XMLName   xml.Name
+	MessageID string     `xml:"message-id,attr"`
+	Request   RPCRequest `xml:",any"`
+}
+
+// RPCRequest describes an RPC request.
+type RPCRequest struct {
+	XMLName xml.Name
+	Body    string `xml:",innerxml"`
+}
+
+// RpcReplyMessage  and ReplyData represent an rpc-reply message that will be sent to a client session, where the
+// element type of the reply body (i.e. the content of the data element)
+// is unknown.
+type RpcReplyMessage struct {
+	XMLName   xml.Name          `xml:"urn:ietf:params:xml:ns:netconf:base:1.0 rpc-reply"`
+	Errors    []common.RPCError `xml:"rpc-error,omitempty"`
+	Data      ReplyData         `xml:"data"`
+	Ok        bool              `xml:",omitempty"`
+	RawReply  string            `xml:"-"`
+	MessageID string            `xml:"message-id,attr"`
+}
+type ReplyData struct {
+	XMLName xml.Name `xml:"data"`
+	Data    string   `xml:",innerxml"`
+}
+
+// RequestHandler is a function type that will be invoked by the session handler to handle an RPC
+// request.
+type RequestHandler func(h *SessionHandler, req *RpcRequestMessage)
+
+// NewServer creates a new Server that will accept Netconf localhost connections on an ephemeral port (available
+// via Port(), with credentials defined by TestUserName and TestPassword.
+// tctx will be used for handling failures; if the supplied value is nil, a default test context will be used.
```

Comment:
tctx -> ctx

Reply:
Have removed this part of the comment (copied from test server) - does not apply here